nixos/modules/services/x11/display-managers/set-session.py

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python
+
+import gi, argparse, os, logging, sys
+gi.require_version("AccountsService", "1.0")
+from gi.repository import AccountsService, GLib
+
+def get_session_file(session):
+    system_data_dirs = GLib.get_system_data_dirs()
+
+    session_dirs = []
+    for data_dir in system_data_dirs:
+        dir = os.path.join(data_dir, "wayland-sessions")
+        session_dirs.append(dir)
+        dir = os.path.join(data_dir, "xsessions")
+        session_dirs.append(dir)
+
+    session_files = []
+    for dir in session_dirs:
+        built_path = os.path.join(dir, session + ".desktop")
+        if os.path.exists(built_path):
+            session_files.append(built_path)
+
+    session_files = list(dict.fromkeys(session_files))  # Remove duplicates
+
+    # Sorta deal with duplicate wayland-sessions and xsessions.
+    # Needed for the situation in gnome-session.
+    for i, file in enumerate(session_files):
+        if len(session_files) < 2:
+            break
+        else:
+            if "/xsessions/" in file:
+                session_files.remove(file)
+
+    if len(session_files) < 1:
+        logging.warning("No session files are found.")
+        sys.exit(0)
+    else:
+        return session_files[0]
+
+
+def is_session_xsession(session_file):
+    if "/xsessions/" in session_file:
+        return True
+    else:
+        return False
+
+
+def is_session_wayland(session_file):
+    if "/wayland-sessions/" in session_file:
+        return True
+    else:
+        return False
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="Set session type for all normal users."
+    )
+    parser.add_argument("session", help="Name of session to set.")
+
+    args = parser.parse_args()
+
+    session = getattr(args, "session")
+    session_file = get_session_file(session)
+
+    user_manager = AccountsService.UserManager.get_default()
+    users = user_manager.list_users()
+
+    for user in users:
+        if user.is_system_account():
+            break
+        else:
+            if is_session_wayland(session_file):
+                logging.debug(
+                    f"Setting session name: {session}, as we found the existing wayland-session: {session_file}"
+                )
+                user.set_session(session)
+            elif is_session_xsession(session_file):
+                logging.debug(
+                    f"Setting session name: {session}, as we found the existing xsession: {session_file}"
+                )
+                user.set_x_session(session)
+            else:
+                raise Exception(f"Couldn't figure out session type for {session_file}")
```

Comment:
Perhaps `sys.exit(1)` would be cleaner than throwing exception.

Reply:
`sys.exit` does not convey any semantic meaning of why the failure happened, it also doesn't provide a stack trace.
In this case I think a `raise` is much better.

Reply:
@adisbladis Same. I was thinking if I ever had to debug this I'd want warnings, exceptions, and errors.

Reply:
Why would we need a stack trace for a normal failure? The checks are same for all the users so it will always fail in the first non-system account. I also meant that the exit would be preceded by the `logging.error` or something.

Reply:
How about @jtojnar?
```diff
diff --git a/nixos/modules/services/x11/display-managers/set-session.py b/nixos/modules/services/x11/display-managers/set-session.py
index 1c0810fadbf..7e883372f69 100755
--- a/nixos/modules/services/x11/display-managers/set-session.py
+++ b/nixos/modules/services/x11/display-managers/set-session.py
@@ -78,7 +78,10 @@ def main():
                 )
                 user.set_x_session(session)
             else:
-                raise Exception(f"Couldn't figure out session type for {session_file}")
+                logging.error(
+                    f"Couldn't figure out session type for {session_file}"
+                )
+                sys.exit(1)
 
 
 if __name__ == "__main__":
```

Reply:
Yeah, I would use that.

Reply:
fixed that in 519448d2e3b18969b16d7509269aa5d02ecf3597.